File: qNoise.cpp
/*
qNoise: A generator of non-Gaussian colored noise
Copyright (c) 2015, Juan Ignacio Deza
email: juan.ignacio.deza@upc.edu

Description
qNoise is a non-gaussian colored random noise generator. It is a handy source of self-correlated noise for a great variety of applications. It depends on two parameters only: tau for controlling the autocorrelation, and q for controlling the statistics. This noise tends smoothly  for q = 1 to an  Ornstein-Uhlenbeck (colored gaussian) noise with autocorrelation tau. for q < 1 it is bounded noise and it is supra-Gaussian for q > 1.
The noise is generated  via a stochastic differential equation using the Heun method (a second order Runge-Kutta type integration scheme) and it is implemented as a stand-alone library in c++.
It Useful as input for numerical simulations, as a source of noise for controlling experiments using synthetic noise via micro-controllers and for a wide variety of applications.

Requirements
It is a stand-alone library with no dependencies other than the standard libraries.
Due to it's use of some functions from the <random> library the library currently works on c++11 or higher only.
This should be OK for most Macs and new Linux systems.
In some older systems it is possible that you need to add `-std=gnu++11` to your compilation flags.

Licence
 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions
 are met:

   1. Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

   2. Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.

   3. The names of its contributors may not be used to endorse or promote
      products derived from this software without specific prior written
      permission.

 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 POSSIBILITY OF SUCH DAMAGE.
 */

#include"qNoise.h"

//Manual seeding.

void qNoiseGen::seedManual(unsigned UserSeed) {
    seed = UserSeed;
    generator.seed(seed);
}

//Timer seeding
//Note: when using this on multiple threads, call seedManual on each thread using different values of seed

void qNoiseGen::seedTimer() {
    // obtain a seed from the timer
    myclock::duration d = myclock::now() - beginning;
    seed = d.count();
    generator.seed(seed);

}

//Gaussian White noise

double qNoiseGen::gaussWN() {
    return randNorm(generator);
}

//Ornstein-Uhlembeck noise type.

double qNoiseGen::orsUhl(double eta, double tau, double H) {
    return eta * exp(-H / tau) + sqrt((1 - exp(-2 / tau * H)) / 2 / tau) * randNorm(generator);
}

/*
 * qNoise Potential derivative. Private Function for use only by qNoise
 * The potential is:
 * \frac{\text{sigma}^2 \log \left(\frac{\text{eta}^2 (q-1) \text{tau}}{\text{sigma}^2}+1\right)}{2 (q-1) \text{tau}}
 * it's derivative is:
 * \frac{\text{eta}}{\frac{\text{eta}^2 (q-1) \text{tau}}{\text{sigma}^2}+1}
 * The /tau is the tau present in the differential equation.
 */
double qNoiseGen::potQNoisePrime(double eta, double tau, double q) {
    return (eta / (1 + (eta * eta * tau * (q - 1)))) / tau;
}

/*qNoise.
 * This functions integrates a differential equation using the Heun Method
 * for q=1 it behaves like the orstein Uhlembeck noise
 * for q<1 it it is defined in a acotated support only (+/- etaCut)
 * for q>1 its statistics are more than gaussian tending (supra-gaussian)
 */
double qNoiseGen::qNoise(double eta, double tau, double q, double H, double sqrt_H = -1) {
    double kHeun, lHeun, differential;
    bool error = false;
    int countError = 0;
    //If the square root of H is provided, it will be used, otherwise
    //calculate it every time the function is invoked.
    if (sqrt_H < 0)
        sqrt_H = sqrt(H);
    //The cut value.
    double etaCut = 1 / sqrt(tau * (1 - q));
    while (1) {
        kHeun = H * potQNoisePrime(eta, tau, q);
        lHeun = sqrt_H * randNorm(generator) / tau;
        differential = -H / 2 * (potQNoisePrime(eta, tau, q)
                + potQNoisePrime(eta + kHeun + lHeun, tau, q)) + lHeun;
        /*
         * Check if the system is inside the boundary.
         * This is only important when q<1.
         * If the eta + D_eta are outside the bounds, it retries 10 times.
         * If it is still out of bounds, it retries with an OH noise, another 10 times
         * This very time step will be Gaussian but with the correct tau.
         * If it is still out of bounds it hard resets it with gaussian noise.
         * The bool error is on place to debug these errors.
         * However it should be a concern only for very low values of q and very
         * hgih values of tau.
         */
        if ((fabs(eta + differential) > etaCut) || isnan(eta + differential)) {
            countError++;
            if (countError > 20) {
                return etaCut * randNorm(generator);
                if (error) std::cerr << "Out of bounds phase 3: " << eta << "\t" << differential << std::endl;
            }
            if (countError > 10) {
                eta = etaCut * orsUhl(eta, tau, H);
                if (error) std::cerr << "Out of bounds phase 2: " << eta << "\t" << differential << std::endl;
            } else {
                if (error) std::cerr << "Out of bounds: " << eta << "\t" << differential << std::endl;
            }
        } else {
            return eta + differential;
        }
    }
}

/*qNoiseNorm.
 * This functions works as qNoise but where both tau and the variance of the noise
 * are independent of q to first order.
 * This approximation fails for q -> 5/3, where both variables diverge.
 */
double qNoiseGen::qNoiseNorm(double eta, double tau, double q, double H, double sqrt_H = -1) {
    return qNoise(eta, tau * (5 - 3 * q) / 2, q, H, sqrt_H);
}

File: go/qnoise.go
/*
qNoise: A generator of non-Gaussian colored noise
Copyright (c) 2021-2025 J. Ignacio Deza

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

qNoise - Non-Gaussian colored noise generator
Contact: ignacio.deza@uwe.ac.uk
Paper: https://doi.org/10.1016/j.softx.2022.101034
*/

package qnoise

import (
	"math"
	"math/rand"
	"time"
)

// Generator holds the random number generator state
type Generator struct {
	rng *rand.Rand
}

// NewGenerator creates a new qNoise generator with a time-based seed
func NewGenerator() *Generator {
	return &Generator{
		rng: rand.New(rand.NewSource(time.Now().UnixNano())),
	}
}

// NewGeneratorWithSeed creates a new qNoise generator with a specific seed
func NewGeneratorWithSeed(seed int64) *Generator {
	return &Generator{
		rng: rand.New(rand.NewSource(seed)),
	}
}

// GaussWN generates a single Gaussian white noise value
func (g *Generator) GaussWN() float64 {
	return g.rng.NormFloat64()
}

// OrsUhl generates a single Ornstein-Uhlenbeck noise step
// Parameters:
//   - eta: current noise value
//   - tau: autocorrelation time
//   - h: integration time step
func (g *Generator) OrsUhl(eta, tau, h float64) float64 {
	expFactor := math.Exp(-h / tau)
	return eta*expFactor + math.Sqrt((1-math.Exp(-2*h/tau))/(2*tau))*g.rng.NormFloat64()
}

// potQNoisePrime calculates the derivative of the qNoise potential (private)
func potQNoisePrime(eta, tau, q float64) float64 {
	return (eta / (1 + (eta*eta*tau*(q-1)))) / tau
}

// QNoise generates a single qNoise step using Heun's method
// Parameters:
//   - eta: current noise value
//   - tau: autocorrelation time
//   - q: statistics parameter (q=1: Gaussian, q<1: bounded, q>1: heavy-tailed)
//   - h: integration time step
//   - sqrtH: square root of h (pass -1 to compute automatically)
func (g *Generator) QNoise(eta, tau, q, h, sqrtH float64) float64 {
	if sqrtH < 0 {
		sqrtH = math.Sqrt(h)
	}

	etaCut := 1 / math.Sqrt(tau*(1-q))
	countError := 0

	for {
		kHeun := h * potQNoisePrime(eta, tau, q)
		lHeun := sqrtH * g.rng.NormFloat64() / tau
		differential := -h/2*(potQNoisePrime(eta, tau, q)+potQNoisePrime(eta+kHeun+lHeun, tau, q)) + lHeun

		if math.Abs(eta+differential) > etaCut || math.IsNaN(eta+differential) {
			countError++
			if countError > 20 {
				sign := 1.0
				if eta < 0 {
					sign = -1.0
				}
				return sign * etaCut * (0.9 + 0.1*g.rng.Float64())
			}
			if countError > 10 {
				eta = etaCut * g.OrsUhl(eta, tau, h)
			}
			// continue loop
		} else {
			return eta + differential
		}
	}
}

// QNoiseNorm generates normalized qNoise where tau and variance are independent of q
func (g *Generator) QNoiseNorm(eta, tau, q, h, sqrtH float64) float64 {
	adjustedTau := tau * (5 - 3*q) / 2
	return g.QNoise(eta, adjustedTau, q, h, sqrtH)
}

// Generate generates an array of qNoise values
// Parameters:
//   - tau: autocorrelation time
//   - q: statistics parameter
//   - n: number of samples
//   - h: integration time step (use 0.01*tau as default)
//   - tempN: number of transient samples to skip (-1 for auto: 2*tau/h)
//   - norm: use normalized version
func (g *Generator) Generate(tau, q float64, n int, h float64, tempN int, norm bool) []float64 {
	if tau < 0 {
		tau = -tau
	}
	if tau == 0 {
		tau = h
	}

	if tempN < 0 {
		tempN = int(2 * tau / h)
	}

	sqrtH := math.Sqrt(h)
	x := g.rng.NormFloat64() / 100.0
	result := make([]float64, n)

	// Skip transient
	for i := -tempN; i < 0; i++ {
		if norm {
			x = g.QNoiseNorm(x, tau, q, h, sqrtH)
		} else {
			x = g.QNoise(x, tau, q, h, sqrtH)
		}
	}

	// Generate output
	for i := 0; i < n; i++ {
		if norm {
			x = g.QNoiseNorm(x, tau, q, h, sqrtH)
		} else {
			x = g.QNoise(x, tau, q, h, sqrtH)
		}
		result[i] = x
	}

	return result
}

// GenerateOrnsteinUhlenbeck generates an array of Ornstein-Uhlenbeck noise
func (g *Generator) GenerateOrnsteinUhlenbeck(tau float64, n int, h float64, tempN int, whiteNoise bool) []float64 {
	if tau < 0 {
		tau = -tau
	}
	if tau == 0 || whiteNoise {
		// Generate white noise
		result := make([]float64, n)
		for i := 0; i < n; i++ {
			result[i] = g.GaussWN()
		}
		return result
	}

	if tempN < 0 {
		tempN = int(2 * tau / h)
	}

	x := g.rng.NormFloat64() / 100.0
	result := make([]float64, n)

	// Skip transient
	for i := -tempN; i < 0; i++ {
		x = g.OrsUhl(x, tau, h)
	}

	// Generate output
	for i := 0; i < n; i++ {
		x = g.OrsUhl(x, tau, h)
		result[i] = x
	}

	return result
}

// Package-level convenience functions using a default generator

var defaultGenerator = NewGenerator()

// Generate generates qNoise using the default generator
func Generate(tau, q float64, n int, h float64, tempN int, norm bool) []float64 {
	return defaultGenerator.Generate(tau, q, n, h, tempN, norm)
}

// GenerateOrnsteinUhlenbeck generates Ornstein-Uhlenbeck noise using the default generator
func GenerateOrnsteinUhlenbeck(tau float64, n int, h float64, tempN int, whiteNoise bool) []float64 {
	return defaultGenerator.GenerateOrnsteinUhlenbeck(tau, n, h, tempN, whiteNoise)
}

// SetSeed sets the seed for the default generator
func SetSeed(seed int64) {
	defaultGenerator = NewGeneratorWithSeed(seed)
}